File: src/01_data_structures/05_hash_table/implementations/ts/hashTable.ts
import { HashTableLinkedList } from "./hashTableLinkedList";

interface IHashTable<T> {
  set(key: string, value: T): HashTable<T>;
  get(key: string): T;
  remove(key: string): HashTable<T>;
}

export class HashTable<T> implements IHashTable<T> {
  private buckets: Array<HashTableLinkedList<[string, T]>>;
  private size;

  constructor(size = 127) {
    this.buckets = new Array(size);
    this.size = size;
  }

  private hash(key: string): number {
    const stringifiedKey = String(key);
    let hash = 0;

    for (let i = 0; i < stringifiedKey.length; i++) {
      hash = hash + stringifiedKey.charCodeAt(i);
    }

    return hash % this.size;
  }

  set(key: string, value: T): HashTable<T> {
    const index = this.hash(key);

    if (!this.buckets[index]) {
      this.buckets[index] = new HashTableLinkedList();
    }

    this.buckets[index].append([key, value]);

    return this;
  }

  get(key: string): T | null {
    const index = this.hash(key);
    const bucket = this.buckets[index];
    if (!bucket) return null;

    const item = bucket.getItemByCallback((item) => item[0] === key);
    if (!item) return null;

    return item[1];
  }

  remove(key: string): HashTable<T> {
    const index = this.hash(key);
    const bucket = this.buckets[index];

    if (!bucket) return this;

    bucket.remove((item) => item[0] === key);

    return this;
  }
}

File: src/01_data_structures/05_hash_table/implementations/go/hash_table.go
package hash_table

import (
	linkedList "data-structures-algorithms-problems/src/01_data_structures/02_linked_list/implementations/go"
)

var DEFAULT_HASH_TABLE_SIZE = 50

type IHashTable[T comparable] interface {
	Set(key string, value T) IHashTable[T]
	Get(key string) (T, bool)
	Remove(key string) IHashTable[T]
}

type HashTable[T comparable] struct {
	buckets []*HashTableLinkedList[T]
	size    int
}

func NewHashTable[T comparable](size int) HashTable[T] {
	if size == 0 {
		size = DEFAULT_HASH_TABLE_SIZE
	}

	buckets := make([]*HashTableLinkedList[T], size)

	return HashTable[T]{
		buckets: buckets,
		size:    size,
	}
}

func (hashTable *HashTable[T]) hash(key string) int {
	var hash int

	for _, r := range key {
		hash += int(r)
	}

	result := hash % hashTable.size

	return result
}

func (hashTable HashTable[T]) Set(key string, value T) IHashTable[T] {
	index := hashTable.hash(key)
	bucket := hashTable.buckets[index]

	if bucket == nil {
		hashTable.buckets[index] = &HashTableLinkedList[T]{
			LinkedList: &linkedList.LinkedList[HashTableLinkedListItem[T]]{},
		}

		bucket = hashTable.buckets[index]
	}

	bucket.Append(HashTableLinkedListItem[T]{Key: key, Value: value})

	return hashTable
}

func (hashTable HashTable[T]) Get(key string) (T, bool) {
	var zero T
	index := hashTable.hash(key)
	bucket := hashTable.buckets[index]

	if bucket == nil {
		return zero, false
	}

	var item, ok = bucket.GetItemIf(func(v HashTableLinkedListItem[T]) bool { return v.Key == key })

	if !ok {
		return zero, false
	}

	return item, true
}

func (hashTable HashTable[T]) Remove(key string) IHashTable[T] {
	index := hashTable.hash(key)
	bucket := hashTable.buckets[index]

	if bucket != nil {
		bucket.Remove(func(item HashTableLinkedListItem[T]) bool { return item.Key == key })
	}

	return hashTable
}
